ux: place typeclass instances in companion objects

gvolpe opened the pull request:
Hey @Iltotore , I know you're still thinking about this, but I went ahead and made the necessary changes so you can see how all those `given` imports go away (as the `catsValidation` and `formCats` examples demonstrate).

The following explicit `given` import WAS only necessary for intersection and union types, which are kind of _orphan instances_.

```scala
import io.github.iltotore.iron.{given, *}
```

But I managed to solve this by moving these instances to the `Constraint` companion object, so we don't need any explicit `given` import at all! 🥳 I believe this is a massive improvement in user experience :)

---

This is a common approach in functional programming (not only in Scala!). Typeclass instances need to be declared together with their type declaration. In Scala, this is companion objects. In Haskell, this is the module where the type is defined. 

As mentioned in #67, [Refined does no different](https://github.com/fthomas/refined/blob/master/modules/core/shared/src/main/scala/eu/timepit/refined/char.scala#L28). I think we could take a lot of inspiration from previous work done in the community :)

---

Unrelated: I noticed many files don't have a newline ending, and this could be problematic to Unix users (Linux, Mac, etc), as these are not considered files according to the POSIX standard. Most IDEs and text editors these days allow you to have an "add newline on save" feature, happy to assist with this if you need help.

Iltotore commented:
> which are kind of orphan instances.

Are they're still orphan instances if the macros/given in is the `IsIntersection`/`IsUnion` companion object?

Iltotore commented:
> Most IDEs and text editors these days allow you to have an "add newline on save" feature, happy to assist with this if you need help.

I think Intellij IDEA and VS Code have such feature but I will check if ScalaFMT has a rule for this has it is probably the best, IDE-agnostic solution

Iltotore commented on main/src/io/github/iltotore/iron/macros/package.scala:
```diff
@@ -7,174 +7,171 @@ import scala.quoted.*
  * Internal macros.
  * @see [[compileTime]] for public compile-time utilities
  */
-package object macros:
```
Since it's not a package object anymore, maybe `package.scala` should be renamed? I don't have a better name thought.

gvolpe replied:
I'm in the same boat haha, Scala 3 decided to deprecate package objects, but don't tell us how to rename them now! 😄 I think it's okay to have it named `package.scala` until we get a better name idea. What matters is that a rename won't affect users as long as the package name remains the same.

Iltotore commented on main/src/io/github/iltotore/iron/package.scala:
```diff
@@ -14,126 +14,138 @@ import scala.util.NotGiven
  * - Refining methods (both compile-time and runtime)
  * - Export of [[constraint.any]].*
  */
-package object iron:
```
Same as for `macros/package.scala`

Iltotore commented on main/test/src/io/github/iltotore/iron/testing/package.scala:
```diff
@@ -1,19 +1,20 @@
 import scala.compiletime.constValue
-import io.github.iltotore.iron.compileTime.stringValue
 
-package object testing:
```
Same as for `macros/package.scala`

Iltotore reviewed (changes requested):
Great job! Here are some small things to change.

gvolpe commented on examples/catsValidation/src/io/github/iltotore/iron/catsValidation/Main.scala:
```diff
@@ -3,12 +3,10 @@ package io.github.iltotore.iron.catsValidation
+import io.github.iltotore.iron.*
 import io.github.iltotore.iron.cats.*
-import io.github.iltotore.iron.constraint.numeric.given
+import io.github.iltotore.iron.constraint.collection.Contain
```
That's usually my preference as well 👍🏽 How do you want to address this? Documentation and examples showing this way?

Iltotore replied:
> Documentation and examples showing this way?

I think this is the right thing to do but I think it should not be done in this PR because it's unrelated.

I already have to polish some parts of the documentation so I will do it later.

gvolpe commented on main/src/io/github/iltotore/iron/constraint/all.scala:
```diff
@@ -0,0 +1,6 @@
+package io.github.iltotore.iron.constraint
+
+object all:
```
I had to remove `export collection.*` because the three types in there are already being re-exported in the `string` package. To be honest, I don't feel so sure about that re-export. What do you say we get rid of it in `string`?

Iltotore replied:
> I had to remove export collection.* because the three types in there are already being re-exported in the string

Does the re-export make some issue or "conflict"? I personally don't want to merge `string` and `collection` together because `string` contains some String-specific constraint like `Match`, `UpperCase` etc...

If there's a conflict, then a better workaround IMO would be to move `MinLength` and `MaxLength` givens in `collection`.

gvolpe replied:
> Does the re-export make some issue or "conflict"? 

It doesn't compile 😄 

> I personally don't want to merge string and collection together because string contains some String-specific constraint like Match, UpperCase etc...

I didn't mean to imply that! I think the `string` package should be as it is, but just import the `colllection` types needed instead of doing the re-export of them so we can avoid these conflicts.

gvolpe commented:
> Are they're still orphan instances if the macro/given is in the IsIntersection/IsUnion companion object?

The line gets blurry with macros, I'd need to double-check everything related to it. I stopped when everything started working without importing any given (I needed to leave home too), but there could be some edge case missing...

Iltotore commented:
> > Are they're still orphan instances if the macro/given is in the IsIntersection/IsUnion companion object?
> 
> The line gets blurry with macros, I'd need to double-check everything related to it. I stopped when everything started working without importing any given (I needed to leave home too), but there could be some edge case missing...

No problem. Thank you for the efforts you already put in Iron.

gvolpe commented on main/src/io/github/iltotore/iron/Constraint.scala:
```diff
@@ -11,3 +14,22 @@ trait Constraint[A, C]:
 
   inline def message: String
+
+object Constraint:
```
> Not should probably be in the same file than union and intersection constraints

Can you help me understand the reasoning? Union and intersection types are language primitives, whereas `Not` is just a custom type declared in Iron. 

> Also isn't a constraint.operator(s) package better suited?

What for? Not sure I understand the idea here.

Iltotore replied:
> I wouldn't be opposed to that, but could you help me understand the reasoning? Union and intersection types are language primitives, whereas Not is just a custom type declared in Iron.

I tend to see intersection and union respectively has "and" and "or" constraints (like boolean "or"/"and") so I originally put the three operators or/and/not in the same file.

In fact, intersection and union replaced Iron 1's `Or` and `And`  constraints which were "normal" constraints just like `Not`.

gvolpe replied:
> Ah that makes sense then, didn't know about the previous Or and And constraints!

While it makes sense on paper, it doesn't really fit anywhere else. Intersections and unions are defined in the macros package, and `Not` is not a macro. Where would you place it, if not under `constraint.any`? Under `constraint.operator` as suggested?

Iltotore replied:
I would place the three "operators" (and/or/not) together whether in `Constraint` companion as you already did for and/or (intersection/union) or in another object like `io.github.iltotore.iron.constraint.operator`.

gvolpe replied:
The latter makes more sense, as there are no operators in the `Constraint` object, only "orphan" instances for intersection and union types from the language.

gvolpe commented:
> No problem. Thank you for the efforts you already put in Iron.

All good, thanks for taking the big change so well! I will go through everything once again, and make sure to fine-tune whatever is left. The idea with the initial PR was to share with code the idea I was trying to convey with typeclass instances 💡

gvolpe commented:
@Iltotore I addressed the macros instances stuff, now it looks tidy 👍🏽 I moved the `Implication` type to its own file and the union implication also lives there. All typeclass instances now live within companion objects.

I also updated the imports and some minor stuff in the documentation.

Iltotore commented on main/src/io/github/iltotore/iron/constraint/operators.scala:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Constraint operators (e.g [[operators.Not]]...).
+ */
+object operators:
```
I mean, I think `Not`, union and intersection should be in the same package. I would rather prefer one of these suggestions:
- Move `Not`, union and intersection in `operator`
- Move `Not` with union and intersection in `Constraint`'s companion
- Keep union and intersection in `Constraint`s companion and `Not` in `any`. (Althrough I think the two above are better solutions)

gvolpe replied:
I feel all your points need some clarification 😄 

> Move Not, union and intersection in operator

What do you mean with moving union and intersection? These are language types.

> Move Not with union and intersection in Constraint's companion

This does not make sense to me, as commented [here](https://github.com/Iltotore/iron/pull/70#discussion_r1039509353).

> Keep union and intersection in Constraints companion and Not in any. (Althrough I think the two above are better solutions)

There are only instances under the `Constraint` object, not type definitions, so not sure what you mean.

gvolpe replied:
After re-reading all your suggestions, I think you're confusing customs types (`Not`) with language constructs (`&` and `|`). Here's what would happen if we follow along any of these suggestions. 

> Move Not, union and intersection in operator

If you mean move union and intersection typeclass instances to the `constraint.operator` package, users will need to import `iron.{given, *}` to get union and intersection types working.

> Move Not with union and intersection in Constraint's companion

Users would need to import `iron.Constraint.Not` to use `Not`, losing consistency with other constraints.

> Keep union and intersection in Constraints companion and Not in any. (Althrough I think the two above are better solutions)

This is how it was until the previous suggested change, which I don't think it's that bad, but moving the `Not` type under `constraint.operator` made sense too.

Iltotore replied:
> After re-reading all your suggestions, I think you're confusing customs types (Not) with language constructs (& and |). Here's what would happen if we follow along any of these suggestions.

Actually not confusing but you got my point of view: I tend to prefer to put Not/union/intersection together because the constraints `C1 | C2` behaves like the boolean "or"and the constraint `C1 & C2` behave like "and", just like `Not` is analoguous to the boolean "not".

Yes, internally union and intersection constraints are really different from `Not` but from a user perspective, I think it makes sense to consider them similar.

> If you mean move union and intersection typeclass instances to the constraint.operator package, users will need to import iron.{given, *} to get union and intersection types working.

> Users would need to import iron.Constraint.Not to use Not, losing consistency with other constraints.

Fair point. I didn't think about these two issues (damn orphan instances!). Then it seems justified to keep `Not` in `any` like you originally did.

gvolpe replied:
No problems, I understand how `And`, `Not`, and `Or` are related, but didn't make sense on the implementation side :)

Take `Functor` from Cats, for instance. The type is defined in Cats, and the instances for language constructs `Functor[List]`, `Functor[Option]`, and so on are defined within the `Functor` companion object. 

In Iron's case, it's the `Constraint` type and the instances for language constructs `&` and `|` are defined within the companion object. 

On the other hand, custom types such as `cats.data.Kleisli`, define the `Functor` and other typeclass instances within the `Kleisli` [companion object](https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/data/Kleisli.scala#L162). This is the same case for Iron's `Not`.  

> Hope you understand my point of view now and finally your first arrangement was probably the best sweat_smile

All good :) So what do we do, `any.Not` or `operator.Not`?

Iltotore replied:
I think `any` is fine since we probably wouldnt have a lot of "operators" like `Not`.

Which one do you prefer ?

gvolpe replied:
`any` is fine with me 👍🏽

Iltotore replied:
Then I will merge this PR once the refactor is done. 👍

gvolpe replied:
Done.
